From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 4 Mar 2013 23:46:10 -0500
Subject: [PATCH] Chunk Save Reattempt

We commonly have "Stream Closed" errors on chunk saving, so this code should re-try to save the chunk in the event of failure and hopefully prevent rollbacks.

diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/RegionFile.java b/src/main/java/net/minecraft/world/level/chunk/storage/RegionFile.java
index b5d8cce349a5073d3a9324091f37adfd4821b0d3..47a5ad2866fd96f9ef240b42c19e72e9061a6fe4 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/RegionFile.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/RegionFile.java
@@ -275,7 +275,7 @@ public class RegionFile implements AutoCloseable {
                     return true;
                 }
             } catch (IOException ioexception) {
-                io.papermc.paper.exception.ServerInternalException.reportInternalException(ioexception); // Paper
+                io.papermc.paper.util.SneakyThrow.sneaky(ioexception); // Paper - we want the upper try/catch to retry this
                 return false;
             }
         }
diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileStorage.java b/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
index 18ef7025f7f4dc2a4aff85ca65ff5a2d35a1ef06..6f98292e770fedd18978264045a51454f1947f3c 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
@@ -167,6 +167,7 @@ public class RegionFileStorage implements AutoCloseable {
         }
         // Paper end - rewrite chunk system
         try { // Paper
+        int attempts = 0; Exception laste = null; while (attempts++ < 5) { try { // Paper
 
         if (nbt == null) {
             regionfile.clear(pos);
@@ -191,7 +192,18 @@ public class RegionFileStorage implements AutoCloseable {
                 dataoutputstream.close();
             }
         }
+        // Paper start
+        return;
+        } catch (Exception ex)  {
+            laste = ex;
+        }
+        }
 
+        if (laste != null) {
+            io.papermc.paper.exception.ServerInternalException.reportInternalException(laste);
+            net.minecraft.server.MinecraftServer.LOGGER.error("Failed to save chunk " + pos, laste);
+        }
+        // Paper end
         } finally { // Paper start
             regionfile.fileLock.unlock();
         } // Paper end
